libraries/chain/webassembly/eos-vm-oc/code_cache.cpp

```diff
@@ -0,0 +1,376 @@
+   return nullptr;
+}
+
+code_cache_sync::~code_cache_sync() {
+   //it's exceedingly critical that we wait for the compile monitor to be done with all its work
+   //This is easy in the sync case
+   _compile_monitor_write_socket.shutdown(local::datagram_protocol::socket::shutdown_send);
+   auto [success, message, fds] = read_message_with_fds(_compile_monitor_read_socket);
+   if(success)
+      elog("unexpected response from EOS-VM OC compile monitor during shutdown");
+}
+
+const code_descriptor* const code_cache_sync::get_descriptor_for_code_sync(const digest_type& code_id, const uint8_t& vm_version) {
+   //check for entry in cache
+   code_cache_index::index<by_hash>::type::iterator it = _cache_index.get<by_hash>().find(boost::make_tuple(code_id, vm_version));
+   if(it != _cache_index.get<by_hash>().end()) {
+      _cache_index.relocate(_cache_index.begin(), _cache_index.project<0>(it));
+      return &*it;
+   }
+
+   const code_object* const codeobject = _db.find<code_object,by_code_hash>(boost::make_tuple(code_id, 0, vm_version));
+   if(!codeobject) //should be impossible right?
+      return nullptr;
+
+   std::vector<wrapped_fd> fds_to_pass;
+   fds_to_pass.emplace_back(memfd_for_blob(codeobject->code));
+
+   write_message_with_fds(_compile_monitor_write_socket, compile_wasm_message{ {code_id, vm_version} }, fds_to_pass);
+   auto [success, message, fds] = read_message_with_fds(_compile_monitor_read_socket);
+   EOS_ASSERT(success, wasm_execution_error, "failed to read response from monitor process");
+   EOS_ASSERT(message.contains<wasm_compilation_result_message>(), wasm_execution_error, "unexpected response from monitor process");
+
+   wasm_compilation_result_message result = message.get<wasm_compilation_result_message>();
+   EOS_ASSERT(result.result.contains<code_descriptor>(), wasm_execution_error, "failed to compile wasm");
+
+   check_eviction_threshold(result.cache_free_bytes);
+
+   return &*_cache_index.push_front(std::move(result.result.get<code_descriptor>())).first;
+}
+
+code_cache_base::code_cache_base(const boost::filesystem::path data_dir, const eosvmoc::config& eosvmoc_config, const chainbase::database& db) :
+   _db(db),
+   _cache_file_path(data_dir/"code_cache.bin")
+{
+   static_assert(sizeof(allocator_t) <= header_offset, "header offset intersects with allocator");
+
+   bfs::create_directories(data_dir);
+
+   if(!bfs::exists(_cache_file_path)) {
+      std::ofstream ofs(_cache_file_path.generic_string(), std::ofstream::trunc);
+      EOS_ASSERT(ofs.good(), database_exception, "unable to create EOS-VM Optimized Compiler code cache");
+      bfs::resize_file(_cache_file_path, eosvmoc_config.cache_size);
+      bip::file_mapping creation_mapping(_cache_file_path.generic_string().c_str(), bip::read_write);
+      bip::mapped_region creation_region(creation_mapping, bip::read_write);
+      new (creation_region.get_address()) allocator_t(eosvmoc_config.cache_size, total_header_size);
```

Comment:
The boost::interprocess::rb_tree_best_fit constructor requires enough memory for the header and at least one block.  Is cache_size validated anywhere?